imxrt-hal/src/tempmon.rs

```diff
@@ -0,0 +1,400 @@
+//! # let (_, ipg_hz) = peripherals.ccm.pll1.set_arm_clock(
+//! #    imxrt_hal::ccm::PLL1::ARM_HZ,
+//! #    &mut peripherals.ccm.handle,
+//! #    &mut peripherals.dcdc,
+//! # );
+//! # let mut cfg = peripherals.ccm.perclk.configure(
+//! #    &mut peripherals.ccm.handle,
+//! #    imxrt_hal::ccm::perclk::PODF::DIVIDE_3,
+//! #    imxrt_hal::ccm::perclk::CLKSEL::IPG(ipg_hz),
+//! # );
+//!
+//! // Init temperature monitor with 8Hz measure freq
+//! // 0xffff = 2 Sec. Read more at `measure_freq()`
+//! let mut temp_mon = peripherals.tempmon.init_with_measure_freq(0x1000);
+//! temp_mon.start();
+//!
+//! let mut last_temp = 0_i32;
+//! loop {
+//!     // Get the last temperature read by the measure_freq
+//!     if let Ok(temp) = temp_mon.get_temp() {
+//!         if last_temp != temp {
+//!             // temperature changed
+//!             last_temp = temp;
+//!         }
+//!         // do something else
+//!     }
+//! }
+//! ```
+//!
+//! ## Example 3
+//!
+//! Low and high temperature Interrupt
+//!
+//! enables interrupts for the selected alarm
+//!
+//! NOTE: Make sure that you implemented the TEMP_LOW_HIGH or TEMP_PANIC interrupt handler
+//!
+//! NOTE: TEMP_LOW_HIGH is triggered for `TempSensor low` and `TempSensor high`
+//!
+//! ```no_run
+//! use imxrt_hal::{self, tempmon::TempMon};
+//! use imxrt_hal::ral::interrupt;
+//!
+//! # let mut peripherals = imxrt_hal::Peripherals::take().unwrap();
+//! # let (_, ipg_hz) = peripherals.ccm.pll1.set_arm_clock(
+//! #    imxrt_hal::ccm::PLL1::ARM_HZ,
+//! #    &mut peripherals.ccm.handle,
+//! #    &mut peripherals.dcdc,
+//! # );
+//! # let mut cfg = peripherals.ccm.perclk.configure(
+//! #    &mut peripherals.ccm.handle,
+//! #    imxrt_hal::ccm::perclk::PODF::DIVIDE_3,
+//! #    imxrt_hal::ccm::perclk::CLKSEL::IPG(ipg_hz),
+//! # );
+//!
+//! // init temperature monitor with 8Hz measure freq
+//! // 0xffff = 2 Sec. Read more at `measure_freq()`
+//! let mut temp_mon = peripherals.tempmon.init_with_measure_freq(0x1000);
+//!
+//! // set low_alarm, high_alarm, and panic_alarm
+//! temp_mon.set_alarm_values(-5_000, 65_000, 95_000);
+//!
+//! // use values from registers if you like to compare it somewhere
+//! let (low_alarm, high_alarm, panic_alarm) = temp_mon.alarm_values();
+//!
+//! // enables interrupts for low_high_alarm and not for panic_alarm
+//! unsafe {
+//!     cortex_m::peripheral::NVIC::unmask(interrupt::TEMP_LOW_HIGH);
+//!     cortex_m::peripheral::NVIC::unmask(interrupt::TEMP_PANIC);
+//! }
+//!
+//! // start could fail if the module is not powered up
+//! if temp_mon.start().is_err() {
+//!     temp_mon.power_up();
+//!     temp_mon.start();
+//! }
+//!
+//! #[cortex_m_rt::interrupt]
+//! fn TEMP_LOW_HIGH() {
+//!     // disable the interrupt to avoid endless triggers
+//!     unsafe {
+//!         cortex_m::peripheral::NVIC::mask(interrupt::TEMP_LOW_HIGH);
+//!         cortex_m::peripheral::NVIC::mask(interrupt::TEMP_PANIC);
+//!     }
```

Comment:
> disable the interrupt to avoid endless triggers

Would disabling tempmon measurements, instead of masking interrupts, be another way to (temporarily) disable the interrupts? If that works, should it be the recommended approach (or at least documented as a different approach)?

(I'm thinking about RTIC users with this suggestion. RTIC users may have relinquished interrupt control to the scheduler. They might appreciate docs that remind them disabling measurements will also disable interrupts.)

---

If keeping the `mask` recommendations, know that [`mask` is safe](https://docs.rs/cortex-m/0.7.4/cortex_m/peripheral/struct.NVIC.html#method.mask), so we can drop the `unsafe` block.

```suggestion
//!     // disable the interrupt to avoid endless triggers
//!     cortex_m::peripheral::NVIC::mask(interrupt::TEMP_LOW_HIGH);
//!     cortex_m::peripheral::NVIC::mask(interrupt::TEMP_PANIC);
```